zerver/tornado/event_queue.py

```diff
@@ -838,15 +828,15 @@ def maybe_enqueue_notifications(
     # mention.  Eventually, we'll add settings to allow email
     # notifications to match the model of push notifications
     # above.
-    if idle and (private_message or mentioned or wildcard_mention_notify or stream_email_notify):
-        notice = build_offline_notification(user_profile_id, message_id)
+    if user_data.is_email_notifiable(private_message, sender_id, idle):
+        notice = build_offline_notification(user_data.user_id, message_id)
         if private_message:
             notice["trigger"] = "private_message"
-        elif mentioned:
+        elif user_data.mentioned:
             notice["trigger"] = "mentioned"
-        elif wildcard_mention_notify:
+        elif user_data.wildcard_mention_notify:
             notice["trigger"] = "wildcard_mentioned"
-        elif stream_email_notify:
+        elif user_data.stream_email_notify:
             notice["trigger"] = "stream_email_notify"
         else:
```

Comment:
Reading this presents the opportunity to move this logic into a `user_data.email_notification_trigger()` function, and have this logic be:

```
trigger = user_data.email_notification_trigger()
if trigger is None:
    raise AssertionError(...)
notice = build_offline_notification(user_data.user_id, message_id)
notice["trigger"] = trigger
```

(Effectively, `user_data.is_email_notifiable` could be refactored to be a thin wrapper around the function computing the trigger and then checking if the result is not None)

Reply:
Ok, done